font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,157 @@
 }
 
-impl crate::Rasterize for DirectWriteRasterizer {
-    type Err = Error;
+impl DirectWriteRasterizer {
+    fn rasterize_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        let loaded_font = &self.fonts.get(&glyph.font_key).ok_or(Error::FontNotLoaded)?;
 
-    fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
-        Ok(DirectWriteRasterizer { fonts: Vec::new(), device_pixel_ratio })
+        match self.rasterize_font_char(&loaded_font.font_face, glyph.c, glyph.size) {
+            Ok(glyph) => Ok(glyph),
+            Err(Error::MissingGlyph(_)) => {
+                let fallback_font = self.get_fallback_font(&loaded_font.font, glyph.c)?;
+                self.rasterize_font_char(&fallback_font.create_font_face(), glyph.c, glyph.size)
+            },
+            Err(err) => Err(err),
+        }
     }
 
-    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
-        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+    fn get_fallback_font(&self, font: &Font, c: char) -> Result<Font, Error> {
+        let fallback = match self.font_fallback.as_ref() {
+            Some(fallback) => fallback,
+            None => return Err(Error::MissingGlyph(c)),
+        };
+
+        let utf16_codepoints = encode_utf16(c);
+        let length = utf16_codepoints.len() as u32;
+
+        let locale = get_current_locale();
+        let text_analysis_source = dwrote::TextAnalysisSource::from_text(
+            Box::new(TextAnalysisSourceData { locale, length }),
+            utf16_codepoints,
+        );
+
+        let fallback_result = fallback.map_characters(
+            &text_analysis_source,
+            0,
+            length,
+            &self.font_collection,
+            Some(&font.family_name()),
+            font.weight(),
+            font.style(),
+            font.stretch(),
+        );
+
+        fallback_result.mapped_font.ok_or_else(|| Error::MissingGlyph(c))
+    }
+
+    fn rasterize_font_char(
+        &self,
+        font_face: &FontFace,
+        c: char,
+        size: Size,
+    ) -> Result<RasterizedGlyph, Error> {
+        let offset = GlyphOffset { advanceOffset: 0.0, ascenderOffset: 0.0 };
+
+        let glyph_index = *font_face
+            .get_glyph_indices(&[c as u32])
+            .first()
+            .ok_or_else(|| Error::MissingGlyph(c))?;
+
+        // The DirectWrite documentation states that we should get 0 returned if the glyph
+        // does not exist in the font
+        if glyph_index == 0 {
+            return Err(Error::MissingGlyph(c));
+        }
+
+        let em_size = em_size(size);
+
+        let glyph_run = dwrote::DWRITE_GLYPH_RUN {
+            fontFace: unsafe { font_face.as_ptr() },
+            fontEmSize: em_size,
+            glyphCount: 1,
+            glyphIndices: &(glyph_index),
+            glyphAdvances: &(0.0),
+            glyphOffsets: &(offset),
```

Comment:
What are these parenthesis for?

Reply:
Hangover from the previous author's implementation. I'll remove them